providers/summary/clearlydefined.js

```diff
@@ -179,30 +202,58 @@ class ClearlyDescribedSummarizer {
       manifest.license &&
       SPDX.normalize(typeof manifest.license === 'string' ? manifest.license : manifest.license.type)
     setIfValue(result, 'licensed.declared', license)
+    const registryUrl = buildRegistryUrl(coordinates)
+    const versionUrl = buildVersionUrl(coordinates)
+    const downloadUrl = buildDownloadUrl(coordinates)
+    if (get(result, 'described'))
+      setIfValue(result, 'described.urls', { registry: registryUrl, version: versionUrl, download: downloadUrl })
   }
 
-  addPodData(result, data) {
+  addPodData(result, data, coordinates) {
     setIfValue(result, 'described.releaseDate', extractDate(data.releaseDate))
     setIfValue(result, 'described.projectWebsite', get(data, 'registryData.homepage'))
     const license = get(data, 'registryData.license')
     if (license) {
       setIfValue(result, 'licensed.declared', SPDX.normalize(typeof license === 'string' ? license : license.type))
     }
+
+    const homepage = get(data, 'registryData.homepage')
+    const revision = get(data, 'registryData.sourceInfo.revision')
+
+    const registryUrl = buildRegistryUrl(coordinates)
+    const versionUrl = homepage && revision ? `${homepage}/tree/${revision}` : null
+    const downloadUrl = homepage && revision ? `${homepage}/archive/${revision}.zip` : null
+    if (get(result, 'described'))
+      setIfValue(result, 'described.urls', { registry: registryUrl, version: versionUrl, download: downloadUrl })
   }
 
-  addGemData(result, data) {
+  addGemData(result, data, coordinates) {
     setIfValue(result, 'described.releaseDate', extractDate(data.releaseDate))
     const license = SPDX.normalize(get(data, 'registryData.license'))
     if (license) set(result, 'licensed.declared', license)
     else {
       const licenses = SPDX.normalize((get(data, 'registryData.licenses') || []).join(' OR '))
       setIfValue(result, 'licensed.declared', licenses)
     }
+    const registryUrl = buildRegistryUrl(coordinates)
+    const versionUrl = buildVersionUrl(coordinates)
+    const downloadUrl = buildDownloadUrl(coordinates)
+    if (get(result, 'described'))
+      setIfValue(result, 'described.urls', { registry: registryUrl, version: versionUrl, download: downloadUrl })
   }
 
-  addPyPiData(result, data) {
+  addPyPiData(result, data, coordinates) {
     setIfValue(result, 'described.releaseDate', extractDate(data.releaseDate))
     setIfValue(result, 'licensed.declared', data.declaredLicense)
+    const releases = get(data, 'registryData.releases')
+    let downloadUrl
+    forEach(releases[coordinates.revision], revision => {
+      if (revision.filename.includes('tar.gz')) downloadUrl = revision.url
```

Comment:
Do you want the first or last one?

Reply:
That's a good question.
Looks like that Pypi could have multiple objects for the same release.

For example:

```
"0.7": [
            {
              "comment_text": "",
              "digests": {
                "md5": "235729aa2bf99e1b940eba53fcbfb37c",
                "sha256": "bd053a6d0e5231bec41da6ef524369d64dc9a7f27c5075914ae1b1abc4dd33e5"
              },
              "downloads": -1,
              "filename": "idna-0.7.tar.gz",
              "has_sig": false,
              "md5_digest": "235729aa2bf99e1b940eba53fcbfb37c",
              "packagetype": "sdist",
              "python_version": "source",
              "size": 20294,
              "upload_time": "2014-07-10T03:37:29",
              "url": "https://files.pythonhosted.org/packages/cf/57/53ac056e1a9d028dfce896929d0fc769e53a1e4d0917c2ec163acae4528f/idna-0.7.tar.gz"
            }
          ],
```

has only one object that contains the .tar.gz file URL, in other cases i found multiple objects for the same release, with different file URL and extension.

What do you think could be the best way in this case?

Reply:
No idea unfortunately of how to distinguish between multiple entries for the same component version. Propose the following
* add a TODO comment here noting that we are randomly picking the last (or first) entry that has a URL.
* Open an issue to figure out what the right answer is and in that issue, put some examples of components that have multiple entries for the same version with different URLs

From there we can get some pypi experts to weigh in on the situation and give us advice.

Reply:
Thanks, i will follow your suggestions